### DISCUSSION TITLE
Reuse database connection

### DISCUSSION THREAD

**Discussion by dan-cooke** [remix-run/remix#2198] [Q&A] [answered]
Hi there,

I cannot seem to re-use my mongodb connection between requests.

I have the following connect function:

```ts
import { MongoClient } from 'mongodb';

let mongo: MongoClient;
export const connectToAtlas = async () => {
  if (!process.env.ATLAS_CLUSTER_URI) {
    throw new Error(`Cannot connect to Atlas no URI in env`);
  }
  const client = new MongoClient(process.env.ATLAS_CLUSTER_URI);

  if (!mongo) {
    mongo = await client.connect();
  }

  return mongo.db(`baggers`);
};

```

However `mongo` is always undefined when my loader calls this function.

This is making my app slower than nextjs :(

**Comment by dan-cooke** [comment]
So I followed the example on nextJS https://github.com/vercel/next.js/blob/canary/examples/with-mongodb/lib/mongodb.js

**Reply by kiliman** [reply]
Which adapter are you using? If you're using serverless like Vercel, you can't reuse connections across each request since requests are short-lived and on different processes.

Only adapters that give your a long-running process like Express will support that.

**Reply by dan-cooke** [reply]
I am using fly.io

It seems to support this

**Reply by kiliman** [reply]
Ok, fly.io is a host, and I believe the default template uses Remix App Server (which is a built-in Express server), so my comments still apply.

Anyway, I'm posting an answer.

**Comment by kiliman** [comment, marked as answer]
Remix App Server and Express adapter purge the require cache on every request (only in DEV).

You should store your client reference on the `global` object. Here's how the Prisma client is persisted from the Jokes tutorial.

```ts
let db: PrismaClient;

declare global {
  var __db: PrismaClient | undefined;
}

// this is needed because in development we don't want to restart
// the server with every change, but we want to make sure we don't
// create a new connection to the DB with every change either.
if (process.env.NODE_ENV === "production") {
  db = new PrismaClient();
  db.$connect();
} else {
  if (!global.__db) {
    global.__db = new PrismaClient();
    global.__db.$connect();
  }
  db = global.__db;
}

export { db };
```

**Reply by dan-cooke** [reply]
Awesome yeh, I had looked at the example in the Jokes tutorial, but I had no idea why a `global` was required. ( I still don't really understand that tbh) so I just ignored it.

My bad.

Thanks!

**Reply by MHassanNadeem** [reply]
Here is my code for `MongoClient`, perhaps somebody would find this explicit code for `mongodb` useful.

```ts
import { MongoClient } from "mongodb";
import { config } from "@app/.server/config";

declare global {
  // eslint-disable-next-line no-var
  var __mongoClient: MongoClient | undefined;
}

let mongoClient: MongoClient;

// this is needed because in development we don't want to restart
// the server with every change, but we want to make sure we don't
// create a new connection to the DB with every change either.
if (process.env.NODE_ENV === "production") {
  mongoClient = new MongoClient(config.mongodb.url);
} else {
  if (!global.__mongoClient) {
    global.__mongoClient = new MongoClient(config.mongodb.url);
  }
  mongoClient = global.__mongoClient;
}

export { mongoClient };
```
